### PR TITLE
Add support for Templates in OpenShift recipes

### PR DESCRIPTION
**Pull request by amisevsk**
### What does this PR do?
Add support for Templates as OpenShift recipes, with parameters
- If the parameter does not have a default value, it is substituted with
  an empty string
- If the parameter has a default value, it is substituted when launching
  the workspace
- Generated parameter values are not supported (see https://github.com/fabric8io/kubernetes-client/issues/1340)

This commit does not add dashboard validation (separate PR: https://github.com/eclipse/che/pull/12582), so starting a workspace based off a template requires manually editing an existing workspace config rather than creating a new stack.

The actual change is quite simple: the actual code difference amounts to
```java
clientFactory.create().load(new ByteArrayInputStream(content.getBytes())).get()
```
instead of
```java
clientFactory.create().lists().load(new ByteArrayInputStream(content.getBytes())).get()
```

Sample recipe:
```yaml
kind: Template
apiVersion: v1
metadata:
  name: template-test
objects:
- 
  apiVersion: v1
  kind: Pod
  metadata:
    name: ws
    labels:
      parameter: ${TEST}
  spec:
    containers:
    - 
      image: 'eclipse/che-dev:nightly'
      name: dev
      resources:
        limits:
        memory: 512Mi
parameters:
- name: TEST
  value: test
```

Above recipe formatted for pasting into existing config:
```
kind: Template\napiVersion: v1\nmetadata:\n  name: template-test\nobjects:\n- \n  apiVersion: v1\n  kind: Pod\n  metadata:\n    name: ws\n    labels:\n      parameter: ${TEST}\n  spec:\n    containers:\n    - \n      image: 'eclipse/che-dev:nightly'\n      name: dev\n      resources:\n        limits:\n        memory: 512Mi\nparameters:\n- name: TEST\n  value: test
```

### What issues does this PR fix or reference?
#11504 

#### Release Notes
Support OpenShift templates in workspace stack recipes.

#### Docs PR
Still TODO: docs need to be updated with routes/secrets/templates/etc. notes.

### CODE REVIEW

**Review by sleshchenko — APPROVED** [review]
LGTM. Nice to see templates supported by Kubernetes/OpenShfit infrastructure :+1: 

A small remark to the recipe for testing: indentation is missing before memory field. Resources should be
```
      resources:
        limits:
          memory: 512Mi
```

**Review by skabashnyuk — APPROVED** [review]

**Comment by sleshchenko** [comment]
@amisevsk Note that Kubernetes/OpenShfit recipe is also parsed as the list here https://github.com/eclipse/che/blob/master/wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/DevfileEnvironmentFactory.java#L122
Please check if it can be easily handled there and fix/create a separate issue.

**Review by ibuziuk — APPROVED** [review]

**Comment by amisevsk** [comment]
@sleshchenko I'm not terribly familiar with the devfile codepaths, but it looks like the task would be more complicated there, since it specifically outputs Kubernetes List yaml -- even if we could unmarshal lists/templates, it's not clear how to preserve this when converting back to yaml. I'll create an issue.

**Comment by amisevsk** [comment]
ci-test

**Comment by amisevsk** [comment]
Issue here: https://github.com/eclipse/che/issues/12592

**Review by garagatyi — APPROVED** [review]

**Comment by ibuziuk** [comment]
@vkuznyetsov @rhopp could we please get QA approval before merging to master ?

**Comment by artaleks9** [comment]
Selenium tests execution on Eclipse Che Multiuser on OCP (https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/1511/) doesn't show any regression against this Pull Request.

File: infrastructures/openshift/src/test/java/org/eclipse/che/workspace/infrastructure/openshift/environment/OpenShiftEnvironmentFactoryTest.java
Comment (dmytro-ndp) [inline infrastructures/openshift/src/test/java/org/eclipse/che/workspace/infrastructure/openshift/environment/OpenShiftEnvironmentFactoryTest.java]: looks like a typo

Comment (amisevsk) [reply infrastructures/openshift/src/test/java/org/eclipse/che/workspace/infrastructure/openshift/environment/OpenShiftEnvironmentFactoryTest.java]: Accidental paste -- thanks!

**Comment by riuvshin** [comment]
@dmytro-ndp do you think that comment fix in the unit test requires whole e2e cycle rerun?

**Comment by riuvshin** [comment]
@amisevsk feel free to merge this right away, do not wait for e2e tests report after comment change...

**Comment by dmytro-ndp** [comment]
@riuvshin:
> @dmytro-ndp do you think that comment fix in the unit test requires whole e2e cycle rerun?
No, I don't think so.
I had run e2e tests again because there initial commit was rewritten with new one [later](https://github.com/eclipse/che/pull/12583/commits/26ce0947d9b00de16c55097e0c8fa264e7f6b509).

**Pull request merged**
